doxygen/man/man3/StatisticsPage.3
.TH "StatisticsPage" 3 "Wed Feb 10 2016" "Version 1.0.0.0" "darksilk" \" -*- nroff -*-
.ad l
.nh
.SH NAME
StatisticsPage \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <statisticspage\&.h>\fP
.PP
Inherits QWidget\&.
.SS "Public Slots"

.in +1c
.ti -1c
.RI "void \fBupdateStatistics\fP ()"
.br
.ti -1c
.RI "void \fBupdatePrevious\fP (int, QString, double, double, double, double, QString, int, int)"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBStatisticsPage\fP (QWidget *parent=0)"
.br
.ti -1c
.RI "\fB~StatisticsPage\fP ()"
.br
.ti -1c
.RI "void \fBsetModel\fP (\fBClientModel\fP *model)"
.br
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "int \fBheightPrevious\fP"
.br
.ti -1c
.RI "int \fBconnectionPrevious\fP"
.br
.ti -1c
.RI "int \fBvolumePrevious\fP"
.br
.ti -1c
.RI "double \fBrewardPrevious\fP"
.br
.ti -1c
.RI "double \fBnetPawratePrevious\fP"
.br
.ti -1c
.RI "QString \fBpawratePrevious\fP"
.br
.ti -1c
.RI "double \fBhardnessPrevious\fP"
.br
.ti -1c
.RI "double \fBhardnessPrevious2\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Definition at line 27 of file statisticspage\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "StatisticsPage::StatisticsPage (QWidget * parent = \fC0\fP)\fC [explicit]\fP"

.PP
Definition at line 14 of file statisticspage\&.cpp\&.
.SS "StatisticsPage::~StatisticsPage ()"

.PP
Definition at line 129 of file statisticspage\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "void StatisticsPage::setModel (\fBClientModel\fP * model)"

.PP
Definition at line 123 of file statisticspage\&.cpp\&.
.SS "void StatisticsPage::updatePrevious (int nHeight, QString phase, double nSubsidy, double pHardness, double pHardness2, double pPawrate2, QString Qlpawrate, int peers, int volume)\fC [slot]\fP"

.PP
Definition at line 111 of file statisticspage\&.cpp\&.
.SS "void StatisticsPage::updateStatistics ()\fC [slot]\fP"

.PP
Definition at line 34 of file statisticspage\&.cpp\&.
.SH "Member Data Documentation"
.PP 
.SS "int StatisticsPage::connectionPrevious"

.PP
Definition at line 37 of file statisticspage\&.h\&.
.SS "double StatisticsPage::hardnessPrevious"

.PP
Definition at line 42 of file statisticspage\&.h\&.
.SS "double StatisticsPage::hardnessPrevious2"

.PP
Definition at line 43 of file statisticspage\&.h\&.
.SS "int StatisticsPage::heightPrevious"

.PP
Definition at line 36 of file statisticspage\&.h\&.
.SS "double StatisticsPage::netPawratePrevious"

.PP
Definition at line 40 of file statisticspage\&.h\&.
.SS "QString StatisticsPage::pawratePrevious"

.PP
Definition at line 41 of file statisticspage\&.h\&.
.SS "double StatisticsPage::rewardPrevious"

.PP
Definition at line 39 of file statisticspage\&.h\&.
.SS "int StatisticsPage::volumePrevious"

.PP
Definition at line 38 of file statisticspage\&.h\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for darksilk from the source code\&.


CHANGELOG
Base58 Change, Port Changes and Minor Fixes
Addition of OBJ directory for Daemon builds
Version updated
ReadMe updated
Instamine code for Silk/Dark weaver swap/burn added to main.cpp
OpenSSL update
LevelDB 1.18 Update
Fix to guard against OpenSSL's new strict DER checks, fixed getbalance and other minor fixes
Spelling edits
Enabled locking of Stormnodes via the Stormnode configuration file
Unset Sandstorm checkbox if there were non-anonymized inputs selected in coincontrol, give warning
Removal of target space change and hardfork
Fixed time to switch "out of sync" messages
Fix twice denominatedBalance calculations
Solve performance problems by caching and fix probable security issue
RPC: Allow "sandstorm" only with ENABLE_WALLET
Disconnect Sandstorm timer on Overview page destruction
Fix spaces/monospace fonts in qt console
Mac compatibility so users can switch tabs using CMD+(NUM)
Added better instructions to get libsecp256k1 installed in case of library issues within ReadMe
Added secp256k1 libs to /deps in case for any reason the libs are not available from github anymore
Increased maximum block size to 50MB (50x Bitcoin Core)
Added safe shutdown for Windows
Fixed sync issues and CPU usage
Add event support for Tor
rpc: Prevent easy memory exhaustion attack 
Fixed issue where using setaccount with a foreign address causes the address to be added to your receiving addresses.
Fix priority calculation in CreateTransaction
darksilk.conf added to /src
GUI Styling changes
Statistics Page Added and Icons Updated
Clickable Padlock to Unlock Wallet and Unlock for Staking Only
Fixed GUI bugs
Assertion Error Fix for MacOSX
Sandstorm GUI Fixes
Decentralised DarkSilkMarket added
Overview Transactions UI Fixes
Messaging, GUI Update, Null Pointer Fix, Various Fixes/Changes
Windows Build Fix
Addition of possibly needed dependency
Stormnode Collateral Increase
Marketplace RPC Calls | Cleanup
Amended Spork Max Value for Stormnode Collateral
Removal of unused Spork
Minimum Transaction fee changed to 0.0001DRKSLK
Add support for miniupnpc API version 14
DARKSILK_TIMEDATA_MAX_SAMPLES 200
Add time offset to getpeerinfo output
Don't check duplicate stake while importing
Ignore blocks received while importing
Further Market RPC JSON calls
marketbuyrequest RPC command
marketmybuys RPC call for market
MacOSX Fixes and GUI Fix
Market RPC Minor Fixes
darksilk-qt.pro cleanup
DarkSilk-qt.app renamed to DarkSilk.app for OSX
Replaced stake min age with min confirmation
Limited drop of difficulty per block
Made special marker non spendable
Added nonspendable encoding for block signing key
BIP66 compatibility
Replaced NOP7 with CHECKLOCKTIMEVERIFY
Enforce new script verification flags
Static pos reward
Demand canonical block sig from updated peers
Submitblock: Add submission of stake template
Reward and Min Conf Blocks Changed
Checkpointing Removal
Gitian MacDeploy Graphics Added
IsProtocolV1/IsProtocolV2 hard fork switches removed
FutureDriftV1/FutureDriftV2 reverted back to FutureDrift
CheckStakeKernelHashV1/CheckStakeKernalHashV2 reverted back to CheckStakeKernelHash
Unsigned Int GetTargetSpacing changed to Static Const Int POS_TARGET_SPACING
Fix boost::get usage with Boost 1.58 in rpcrawtransaction.cpp 
Static PoS Reward of 1 DRKSLK
Prevent redefinition compiler warning
Fix issue where getrawtransaction RPC call does not work
PoW additions to GetNextTargetRequired
Scale up addrman
Gitian windows compile fixes
Correctly randomize change output position
LevelDB windows cross compile missing files
Scale up addrman
Gitian windows compile fixes
Correctly randomize change output position
Removal of Dead Code
Fix for building with Boost 1.58
LevelDB Cross compile fixes
DarkSilk-Qt: cleanup / optimise addressbookpage
Discourage fee sniping with nLockTime
Removal of dead code from main.h
Changed ProofOfStakeReward to use static const
Diff changes
Shadowchat Messaging Updates
DNSSeed Thread Fix
Stormnode payments start on block 82002
Restore hex to getrawtransaction vout scriptPubkey
static const rename | nActualSpacing code reform
Removal of Keepass until later date that this doesn't cause an issue with Windows builds
Updated dependencies, added Gitian OSX files
Remove unused GetProofOfStakeReward
Faster Retargetting for PoW
Fixed PoS target spacing
Fix payee selectoin on StormnodePayments::ProcessBlock
fix log output / more notCapableReason-s
HasCollateralInputs should check only for at least 1 collateral not 2
Track lastTimeChanged for mnodeman and ask for update only if UpdateNeeded
Added sstx support for free transactions 
Fix for src/qt/stormnodemanager.cpp
RPC Fixes
Fix DoAutomaticDenominating
Few Fixes for Processing Extra Messages
Reversioning and restriction change
Check and remove expired Stormnodes on Client start
Maps in snodeman
bool warning fix
SSEG Update
Show new number of SN's on add/remove in logs | Move output to debug level
sigTime
Temp Fix prior to major changes
SPORK Additions/Changes | SANDSTORM fee removal
prevent ganing sigTime too much on initial ssee
SandStorm/InstantX Functionality | Send with SandStorm on sendcoinsdialog.ui
Fix Filtering for active | add spaces to table-fy console view | add spaces to IP only in full mode
Merged Stormnode blinding
fix compiling for windows
fix console spam
Make stormnodes.dat verification more specific, prevent from overwriting corruped
files with unknown format, give more info in debug.log output
fixed shared key usage
fix even more console spam
rename stormnodes.dat->mncache.dat / fix strings
save nsqCount in mncache.dat
[Qt] add more NULL pointer checks in darksilkgui.cpp
close wallet regardless of "minimize to tray" settings
Added coincontrol sandstorm rounds column
Stormnode Proof of Service Added
Add structure for donating a percentage of ones earnings to another party via the protocol
Fixed snw message issue
fix pubkey/donation conditions
Fixed blinding submission process
increasing ss timeouts
fixed ssee sync
fixed Stormnode payments FindOldestNotInVec

src/qt/statisticspage.cpp
#include "statisticspage.h"
#include "ui_statisticspage.h"
#include "main.h"
#include "wallet/wallet.h"
#include "init.h"
#include "base58.h"
#include "clientmodel.h"
#include "rpcblockchain.cpp"
#include <sstream>
#include <string>

using namespace json_spirit;

StatisticsPage::StatisticsPage(QWidget *parent) :
    QWidget(parent),
    ui(new Ui::StatisticsPage) {
    ui->setupUi(this);
    setFixedSize(400, 420);

    connect(ui->startButton, SIGNAL(pressed()), this, SLOT(updateStatistics()));
}

int heightPrevious = -1;
int connectionPrevious = -1;
int volumePrevious = -1;
double rewardPrevious = -1;
double netPawratePrevious = -1;
double pawratePrevious = -1;
double hardnessPrevious = -1;
double hardnessPrevious2 = -1;
QString stakecPrevious = "";


void StatisticsPage::updateStatistics() {
    double pHardness = GetDifficulty();
    double pHardness2 = GetDifficulty(GetLastBlockIndex(pindexBest, true));
    int pPawrate = GetPoWMHashPS();
    double pPawrate2 = 0.000;
    int nHeight = pindexBest->nHeight;
    double nSubsidy = 42;
    int64_t volume = ((pindexBest->nMoneySupply) / 100000000);
    int peers = this->model->getNumConnections();
    pPawrate2 = (double)pPawrate;
    ui->progressBar->setValue(nHeight);
    QString height = QString::number(nHeight);
    QString phase = "";

    if (pindexBest->nHeight < 42001) {
        phase = "<p align=\"center\">PoW/PoS</p>";
        ui->progressBar->setMaximum(42001);
    } else {
        ui->progressBar->hide();
        phase = "<p align=\"center\">PoS</p>";
    }

    QString subsidy = QString::number(nSubsidy, 'f', 6);
    QString hardness = QString::number(pHardness, 'f', 6);
    QString hardness2 = QString::number(pHardness2, 'f', 6);
    QString pawrate = QString::number(pPawrate2, 'f', 3);
    QString Qlpawrate = model->getLastBlockDate().toString();

    QString QPeers = QString::number(peers);
    QString qVolume = QLocale(QLocale::English).toString((qlonglong)volume);

    if (nHeight > heightPrevious) {
        ui->progressBar->setValue(nHeight);
        ui->heightBox->setText("<font color=\"green\">" + height + "</font>");
    } else {
        ui->heightBox->setText(height);
    }

    if (phase != stakecPrevious) {
        ui->cBox->setText("<font color=\"green\">" + phase + "</font>");
    } else {
        ui->cBox->setText(phase);
    }

    if (pHardness2 > hardnessPrevious2) {
        ui->diffBox2->setText("<font color=\"green\">" + hardness2 + "</font>");
    } else if (pHardness2 < hardnessPrevious2) {
        ui->diffBox2->setText("<font color=\"red\">" + hardness2 + "</font>");
    } else {
        ui->diffBox2->setText(hardness2);
    }

    if (Qlpawrate != pawratePrevious) {
        ui->localBox->setText("<font color=\"green\">" + Qlpawrate + "</font>");
    } else {
        ui->localBox->setText(Qlpawrate);
    }

    if (peers > connectionPrevious) {
        ui->connectionBox->setText("<font color=\"green\">" + QPeers + "</font>");
    } else if (peers < connectionPrevious) {
        ui->connectionBox->setText("<font color=\"red\">" + QPeers + "</font>");
    } else {
        ui->connectionBox->setText(QPeers);
    }

    if (volume > volumePrevious) {
        ui->volumeBox->setText("<font color=\"green\">" + qVolume + " DRKSLK" + "</font>");
    } else if (volume < volumePrevious) {
        ui->volumeBox->setText("<font color=\"red\">" + qVolume + " DRKSLK" + "</font>");
    } else {
        ui->volumeBox->setText(qVolume + " DRKSLK");
    }

    updatePrevious(nHeight, phase, nSubsidy, pHardness, pHardness2, pPawrate2, Qlpawrate, peers, volume);
}

void StatisticsPage::updatePrevious(int nHeight, QString phase, double nSubsidy, double pHardness, double pHardness2, double pPawrate2, QString Qlpawrate, int peers, int volume) {
    heightPrevious = nHeight;
    stakecPrevious = phase;
    rewardPrevious = nSubsidy;
    hardnessPrevious = pHardness;
    hardnessPrevious2 = pHardness2;
    netPawratePrevious = pPawrate2;
    pawratePrevious = Qlpawrate;
    connectionPrevious = peers;
    volumePrevious = volume;
}

void StatisticsPage::setModel(ClientModel *model) {
    updateStatistics();
    this->model = model;
}


StatisticsPage::~StatisticsPage() {
    delete ui;
}
